### DISCUSSION TITLE
Fast homepage load

### DISCUSSION THREAD

**Discussion by denes44** [Blackymas/NSPanel_HA_Blueprint#733] [Ideas]
Hi,

I read that there are works in progress, to make the page loads faster, but I think it's worth considering making the elements on the homepage global, so they are don't need an update when the user loads the homepage.
Right now for me the loading time is about 1.5-2 sec, plus ~0.5 sec for the display to load the elements.

One feature to which it is essential (in my opinion) is when you use sleep, and wants to wake the display with motion sensors.
The fastest way to display the homepage content after waking up the display from sleep is if the values of the homepage elements are already there, and don't need updating.

**Comment by deejaybeam** [comment]
You are right, but in the past we need to change from global variables to local variables, because otherwise tft does not compile anymore because out-of-memory...
This is just a very limited device, so only option would be to e.g. remove the QR-Code, this takes very much mem

**Reply by denes44** [reply]
Yes, I read that it was globals before, but you needed to change it back because memory reasons.
I suggest to change only the homepage variables back to global.

This is the memory usage of the pages when the variables on the home page are local:
```
Page:home Memory Occupied:216+1044=1260
Page:weather01 Memory Occupied:216+696=912
Page:weather02 Memory Occupied:216+696=912
Page:weather03 Memory Occupied:216+696=912
Page:weather04 Memory Occupied:216+696=912
Page:weather05 Memory Occupied:216+696=912
Page:climate Memory Occupied:216+524=740
Page:settings Memory Occupied:216+412=628
Page:boot Memory Occupied:216+560=776
Page:screensaver Memory Occupied:216+304=520
Page:lightsettings Memory Occupied:216+676=892
Page:coversettings Memory Occupied:216+556=772
Page:buttonpage01 Memory Occupied:216+868=1084
Page:buttonpage02 Memory Occupied:216+868=1084
Page:buttonpage03 Memory Occupied:216+868=1084
Page:buttonpage04 Memory Occupied:216+868=1084
Page:notification Memory Occupied:216+472=688
Page:qrcode Memory Occupied:216+416=632
Page:entitypage01 Memory Occupied:216+916=1132
Page:entitypage02 Memory Occupied:216+916=1132
Page:entitypage03 Memory Occupied:216+916=1132
Page:entitypage04 Memory Occupied:216+916=1132
```
And this is when they are global:
```
Page:home Memory Occupied:900+360=1260
Page:weather01 Memory Occupied:900+696=1596
Page:weather02 Memory Occupied:900+696=1596
Page:weather03 Memory Occupied:900+696=1596
Page:weather04 Memory Occupied:900+696=1596
Page:weather05 Memory Occupied:900+696=1596
Page:climate Memory Occupied:900+524=1424
Page:settings Memory Occupied:900+412=1312
Page:boot Memory Occupied:900+560=1460
Page:screensaver Memory Occupied:900+304=1204
Page:lightsettings Memory Occupied:900+676=1576
Page:coversettings Memory Occupied:900+556=1456
Page:buttonpage01 Memory Occupied:900+868=1768
Page:buttonpage02 Memory Occupied:900+868=1768
Page:buttonpage03 Memory Occupied:900+868=1768
Page:buttonpage04 Memory Occupied:900+868=1768
Page:notification Memory Occupied:900+472=1372
Page:qrcode Memory Occupied:900+416=1316
Page:entitypage01 Memory Occupied:900+916=1816
Page:entitypage02 Memory Occupied:900+916=1816
Page:entitypage03 Memory Occupied:900+916=1816
Page:entitypage04 Memory Occupied:900+916=1816
```

The available memory is 3584, so even the biggest pages are almost half the size this. So nothing needs to be removed for this, and possibly not in the future as well.

**Comment by deejaybeam** [comment]
Maybe you could send also a list of what you have changed from local to gloabal and @Blackymas should have a look - he is the "masterbrain" of the HMI.
In addition this is only the half way - this also requires a big change of the blueprint itself:
Changed entities (like weather, chips, ...) have to be updated also if not on home-page,
And the loading sequence of the home-page needs to be adopted as well

but for the home-page this may makes sense, can´t 100% remember why we also have done this for home-page as well

**Reply by denes44** [reply]
Well, I changed everyting that is on the device canvas. So the visual components only.
As for the blueprint, I already did this for myself in the 3.2 version, and it wasn't a big deal. I can make a PR for it, if needed.

**Reply by deejaybeam** [reply]
> Well, I changed everyting that is on the device canvas. So the visual components only. As for the blueprint, I already did this for myself in the 3.2 version, and it wasn't a big deal. I can make a PR for it, if needed.

but you have to use dev for pull requestst and there has many changed
we are finishing v3.3

**Reply by deejaybeam** [reply]
> > Well, I changed everyting that is on the device canvas. So the visual components only. As for the blueprint, I already did this for myself in the 3.2 version, and it wasn't a big deal. I can make a PR for it, if needed.
> 
> but you have to use dev for pull requestst and there has many changed we are finishing v3.3

see also: https://github.com/Blackymas/NSPanel_HA_Blueprint/blob/dev/CONTRIBUTING.md

**Reply by denes44** [reply]
Yes, I'm aware. I just didn't do it yet, because probably these kind of changed requires approval.
Also what I don't know is how you handle editing the .HMI file. Because that file can't really supports if two persons are working on it simultaneously.

But basically what I did:
- Remove vis 255,0 command from Nextion home page preinit event
- Remove show_all command from blueprint at the end of the big home page update script, because it is not needed
- Modify the visual components to global
- Modify the trigger conditions for the homepage elements to exclude the current page=home page condition

**Reply by deejaybeam** [reply]
yes HMI is tricky - main problem is to see what is changed.
in the dev-repo I have used https://github.com/MMMZZZZ/Nextion2Text to make changes in HMI "visible"
you can see the changes in the subfolders nspanel_eu_code and nspanel_us_code

**Reply by denes44** [reply]
That is a good idea. I tried it in a PR including HMI edit in #740.
I also remembered that you can export a page from one HMI file and you can import it to another. That way you can do a manual merge if two people worked on different pages.

**Reply by edwardtfn** [reply]
> * Remove show_all command from blueprint at the end of the big home page update script, because it is not needed

Please be aware the this `show_all` behavior was changed in #731, so you probably will have to remove the new input entry also.
And `show_all` is also used for button pages.

> * Modify the trigger conditions for the homepage elements to exclude the current page=home page condition

I think this makes sense for some elements with lower rate of change, like date/time, weather condition, temperatures, buttons, etc., but the values could have changes every second and I don't think it would bring a real benefit to update those while the page is not visible. Users will probably be OK about having a obsolete value for the fraction of time the page will be loaded (and those could be moved to be the first thing to be updated when home page opens).
